tikv-client/src/main/java/com/pingcap/tikv/txn/LockResolverClient.java

```diff
@@ -0,0 +1,288 @@
+      }
+    } finally {
+      readWriteLock.writeLock().unlock();
+    }
+  }
+
+  private Long getResolved(Long txnID) {
+    try {
+      readWriteLock.readLock().lock();
+      return resolved.get(txnID);
+    } finally {
+      readWriteLock.readLock().unlock();
+    }
+  }
+
+  public Long getTxnStatus(BackOffer bo, Long txnID, ByteString primary) {
+    Long status = getResolved(txnID);
+
+    if (status != null) {
+      return status;
+    }
+
+    while (true) {
+      // refresh region
+      region = session.getRegionManager().getRegionByKey(primary);
+
+      Supplier<CleanupRequest> factory = () ->
+          CleanupRequest.newBuilder().setContext(region.getContext()).setKey(primary).setStartVersion(txnID).build();
+      KVErrorHandler<CleanupResponse> handler =
+          new KVErrorHandler<>(
+              session.getRegionManager(), this, region, resp -> resp.hasRegionError() ? resp.getRegionError() : null);
+
+      CleanupResponse resp = callWithRetry(bo, TikvGrpc.METHOD_KV_CLEANUP, factory, handler);
+
+      status = 0L;
+      if (resp.hasRegionError()) {
+        bo.doBackOff(BoRegionMiss, new RegionException(resp.getRegionError()));
+        continue;
+      }
+
+      if (resp.hasError()) {
+        logger.error(String.format("unexpected cleanup err: %s, tid: %d", resp.getError(), txnID));
+        throw new KeyException(resp.getError());
+      }
+
+      if (resp.getCommitVersion() != 0) {
+        status = resp.getCommitVersion();
+      }
+
+      saveResolved(txnID, status);
+      return status;
+    }
+  }
+
+  // ResolveLocks tries to resolve Locks. The resolving process is in 3 steps:
+  // 1) Use the `lockTTL` to pick up all expired locks. Only locks that are old
+  //    enough are considered orphan locks and will be handled later. If all locks
+  //    are expired then all locks will be resolved so true will be returned, otherwise
+  //    caller should sleep a while before retry.
+  // 2) For each lock, query the primary key to get txn(which left the lock)'s
+  //    commit status.
+  // 3) Send `ResolveLock` cmd to the lock's region to resolve all locks belong to
+  //    the same transaction.
+  public boolean resolveLocks(BackOffer bo, List<Lock> locks) {
+    if (locks.size() == 0) {
+      return true;
+    }
+
+    List<Lock> expiredLocks = new ArrayList<>();
+    for (Lock lock: locks) {
+      if (TsoUtils.isExpired(lock.getTxnID(), lock.getTtl())) {
+        expiredLocks.add(lock);
+      }
+    }
+
+    if (expiredLocks.size() == 0) {
+      return false;
+    }
+
+    // TxnID -> []Region, record resolved Regions.
+    // TODO: Maybe put it in all LockResolverClient and share by all txns.
+    Map<Long, Set<RegionVerID>> cleanTxns = new HashMap<>();
+    for (Lock l: expiredLocks) {
+      Long status = getTxnStatus(bo, l.getTxnID(), l.getPrimary());
+
+      cleanTxns.putIfAbsent(l.getTxnID(), new HashSet<>());
+      Set<RegionVerID> cleanRegion = cleanTxns.get(l.getTxnID());
```

Comment:
```suggestion
      Set<RegionVerID> cleanRegion = cleanTxns.computeIfAbsent(l.getTxnID(), k -> new HashSet<>());
```
and delete line 183.